### PR TITLE
Easy report generation

### PR DESCRIPTION
**Pull request by karanjitsingh**
- Expose `ReportConfigurationBuilder`
  - Refactor code to be able to create `ReportConfiguration` from a dictionary
- Null check for `RiskHotspotAnalysisResult` in `ReportBuilderBase`
- Expose `ReportContext` `ReportGenerator` `CachingFileReader` `IFileReader`

Creating reports using the SDK becomes as simple as:

```csharp
using ...

namespace htmlreport
{
    class Program
    {
        static void Main(string[] args)
        {
            var result = ParseCoverageFiles(new List<string>() { @"D:\CodeCoverageReport_8440364\coverage.xml" });
            CreateHTMLReport(result, DateTime.Now);
        }

        private static ParserResult ParseCoverageFiles(List<string> coverageFiles)
        {
            CoverageReportParser parser = new CoverageReportParser(
                1,
                new string[] { },
                new DefaultFilter(new string[] { }),
                new DefaultFilter(new string[] { }),
                new DefaultFilter(new string[] { }));

            ReadOnlyCollection<string> collection = new ReadOnlyCollection<string>(coverageFiles);
            return parser.ParseFiles(collection);
        }

        private static void CreateHTMLReport(ParserResult parserResult, DateTime executionTime)
        {
            var summaryResult = new SummaryResult(parserResult);
            var builder = new HtmlInlineAzurePipelinesReportBuilder();

            var folder = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
            Directory.CreateDirectory(folder);

            var config = new ReportConfigurationBuilder().Create(new Dictionary<string, string>() {
                { "targetdir", folder },
                { "reporttypes", "HtmlInline_AzurePipelines" }
            });

            var reportContext = new ReportContext(config, new Settings());

            builder.ReportContext = reportContext;

            new ReportGenerator(
                new CachingFileReader(reportContext.Settings.CachingDuringOfRemoteFilesInMinutes),
                parserResult,
                new List<IReportBuilder>() { builder } )
                    .CreateReport(false, null, executionTime, reportContext.ReportConfiguration.Tag);
        }
    }
}
```

### CODE REVIEW

File: src/ReportGenerator.Core/Reporting/Builders/ReportBuilderBase.cs [hunk 1]
```diff
@@ -237,32 +237,35 @@ public virtual void CreateSummaryReport(IReportRenderer reportRenderer, SummaryR
                 reportRenderer.MetricsTable(new[] { methodMetric });
             }
 
-            if (reportRenderer.SupportsRiskHotsSpots
-                && this.ReportContext.RiskHotspotAnalysisResult.CodeCodeQualityMetricsAvailable)
+            if (this.ReportContext.RiskHotspotAnalysisResult != null)
```
Comment (karanjitsingh) [inline src/ReportGenerator.Core/Reporting/Builders/ReportBuilderBase.cs]: Gtihub diff seems weird

image

**Comment by danielpalme** [comment]
I think you can achieve the same result, with much less code and fewer changes in _ReportGenerator_.
Only the changes to [ReportConfigurationBuilder](https://github.com/danielpalme/ReportGenerator/pull/247/files#diff-805b11466d0a15c11c6601d659e815f5) are required:

The following sample demonstrates how to use the higher level APIs:

```csharp
using ...

namespace htmlreport
{
    class Program
    {
        static void Main(string[] args)
        {
            CreateHTMLReport(new List<string>() { @"D:\CodeCoverageReport_8440364\coverage.xml" });
        }

        private static void CreateHTMLReport(IEnumerable<string> coverageFiles)
        {
            var folder = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
            Directory.CreateDirectory(folder);

            var reportConfigurationBuilder = new ReportConfigurationBuilder();
            ReportConfiguration configuration = reportConfigurationBuilder.Create(new Dictionary<string, string>() {
                { "reports", string.Join(";", coverageFiles) }
                { "targetdir", folder },
                { "reporttypes", "HtmlInline_AzurePipelines" }
            });

            new Generator().GenerateReport(configuration);
        }
    }
}
```

What do you think?

**Comment by karanjitsingh** [comment]
My original idea was to modify Generator.cs itself. I wanted to separate parsing and report creation for scenarios where only parsing is required.

**Comment by danielpalme** [comment]
My recommendation is to:
- Use [CoverageReportParser](https://github.com/danielpalme/ReportGenerator/blob/master/src/ReportGenerator.Core/Parser/CoverageReportParser.cs) to parse coverage files and then do you own processing based on the _ParserResult_.
- Use the approach from [above](https://github.com/danielpalme/ReportGenerator/pull/247#issuecomment-495765295) to generate HTML reports.

Don't use the (currently internal classes). If they become public and you start using them, it would become very hard for me to change the internal behavior without breaking your code.

**Comment by karanjitsingh** [comment]
@danielpalme, is there a way to generate HTML reports based on the `ParserResult` that I have? For example if I simply go through `Generator` it will parse the files for me instead of using my `ParserResult` instance.

My ask here is to decouple these two operations.

I can think of two ways of doing this without exposing internal classes,
- Separate Parsing and report generation in Generator itself, and `GenerateReport` calls into both of these to continue working the same way. This way Generator can expose methods for decoupled parsing and report generation as well.
- Another way of doing this is by having an overloaded implementation of `GenerateReport` to accept an instance of `ParserResult` and use that instead of parsing on it's own. (I personally prefer this one)

It's possible in my scenario that first we parse the files and decide later on in time whether to create HTML reports or not, hence decoupling the two operations would make doing that simpler.

Let me know if this makes sense and I can raise a PR.

**Comment by danielpalme** [comment]
I will have an look at this tomorrow.

**Comment by danielpalme** [comment]
I added a new overload of `GenerateReport`:
https://github.com/danielpalme/ReportGenerator/blob/7f46d553356a9cfa609ab6445b5c06dac282cd3c/src/ReportGenerator.Core/Generator.cs#L162

This (pre-)release contains the new method: 
https://www.nuget.org/packages/ReportGenerator.Core/4.1.9-rc1

**Pull request merged**